routing/osrm_path_segment_factory.cpp

```diff
@@ -143,40 +144,77 @@ void OsrmPathSegmentFactory(RoutingMapping & mapping, Index const & index, RawPa
-  if (isStartNode && isEndNode)
+  size_t const startIndex = isStartNode ? findIntersectingSeg(startGraphNode.segment) : 0;
+  size_t const endIndex = isEndNode ? findIntersectingSeg(endGraphNode.segment) + 1 : buffer.size();
+  // @TODO(bykoianko) When start and end points belong to the same NodeID it's possible
+  // to get endIndex <= startIndex. Production version returns route building error.
+  // Debug version crashes with an assert. The test to show it USALosAnglesAriaTwentyninePalmsHighwayTimeTest.
+  LoadPathGeometry(buffer, startIndex, endIndex, index, mapping, startGraphNode, endGraphNode, isStartNode,
+                   isEndNode, loadedPathSegment);
+
+  LoadedPathSegment fullPathSegment;
+  LoadPathGeometry(buffer, 0, buffer.size(), index, mapping, FeatureGraphNode(), FeatureGraphNode(),
+                   false /* isStartNode */, false /* isEndNode */, fullPathSegment);
+  double const fullPathLengthM = PathLengthM(fullPathSegment.m_path);
+  if (fullPathLengthM == 0.0 || loadedPathSegment.m_path.empty())
+  {
+    loadedPathSegment.m_weight = 0.0;
+    return;
+  }
+
+  // Full weight (time needed to path NodeID).
+  double fullWeight = 0; /* |fullWeight * kOSRMWeightToSecondsMultiplier| is time in seconds. */
+  if (isStartNode)
+  {
+    // Note. It's possible that isStartNode == true and isEndNode == true.
+    // In that case it's ok to follow this branch of if because in that case
+    // startGraphNode.node.forward_node_id == endGraphNode.node.forward_node_id.
+    ASSERT(isStartNode != isEndNode
+           || startGraphNode.node.forward_node_id == endGraphNode.node.forward_node_id
+           || startGraphNode.node.forward_node_id == endGraphNode.node.reverse_node_id, ());
+    fullWeight = (osrmPathSegment.node == startGraphNode.node.forward_node_id)
+        ? startGraphNode.node.forward_offset
+        : startGraphNode.node.reverse_offset;
+  }
+  else /* isEndNode */
   {
-    double const forwardWeight = (osrmPathSegment.node == startGraphNode.node.forward_node_id)
-                                     ? startGraphNode.node.forward_weight
-                                     : startGraphNode.node.reverse_weight;
-    double const backwardWeight = (osrmPathSegment.node == endGraphNode.node.forward_node_id)
-                                      ? endGraphNode.node.forward_weight
-                                      : endGraphNode.node.reverse_weight;
-    double const wholeWeight = (osrmPathSegment.node == startGraphNode.node.forward_node_id)
-                                   ? startGraphNode.node.forward_offset
-                                   : startGraphNode.node.reverse_offset;
-    // Sum because weights in forward/backward_weight fields are negative. Look osrm_helpers for
-    // more info.
-    loadedPathSegment.m_weight = wholeWeight + forwardWeight + backwardWeight;
+    fullWeight = (osrmPathSegment.node == endGraphNode.node.forward_node_id)
```

Comment:
> So, here you put a whole edge weight (time * 10) into the fullWeight variable.

Yes.

> But you need only a part of this weight. You need time from the start of the node to the end point.

Please have a look at the code below. The key line in it below:
```c++
...
  loadedPathSegment.m_weight =
      kOSRMWeightToSecondsMultiplier * fullWeight * partPathLengthM / fullPathLengthM;
...
```

The idea behind it:
We know the length in meters and weight in seconds of OSRM node id for every direction. So we know estimated speed (estimated average) of OSRM node id for every direction. Besides that we know the length of path (start - end_point) and the direction. This way we can calculate loadedPathSegment.m_weight using formula above.

But what's the problem with the former way of loadedPathSegment.m_weight calculation. There're two:
1. I found a place where startGraphNode.node.forward_weight /startGraphNode.node.reverse_weight/ endGraphNode.node.forward_weight / endGraphNode.node.reverse_weight behaves in unpredictable way. (I can write further details if you're interested in knowing).

2. Here (in this method) I can calculate the length of intersection of the navigation route and the NodeID (for start and for end) correctly. But I cannot calculate here correctly the weight based on start(end)GraphNode.node.reverse(forward)_weight(offset). If you need any further details please let me know.